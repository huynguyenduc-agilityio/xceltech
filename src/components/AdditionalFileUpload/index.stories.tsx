import { Meta, StoryObj } from '@storybook/react';

// Components
import AdditionalFileUpload from '.';

const meta = {
  title: 'Components/AdditionalFileUpload',
  component: AdditionalFileUpload,
  parameters: {
    layout: 'centered',
  },
  render: (arg) => (
    <div className="flex justify-center w-[1000px] p-9 bg-white">
      <AdditionalFileUpload {...arg} />
    </div>
  ),
  tags: ['autodocs'],
} satisfies Meta<typeof AdditionalFileUpload>;

export default meta;

type Story = StoryObj<typeof AdditionalFileUpload>;

export const Default: Story = {
  args: {
    name: 'test',
    onFileChange: () => {},
  },
};
