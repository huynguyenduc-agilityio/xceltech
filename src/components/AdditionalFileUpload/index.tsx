import { useCallback, useState } from 'react';
import { useForm } from 'react-hook-form';

import {
  Button,
  FormField,
  FormItem,
  FormMessage,
  Input,
  Label,
} from '../common';

export interface AdditionalFileUploadProps {
  name: string;
  fileUrl?: string;
  onFileChange: (file: string) => void;
}

const AdditionalFileUpload = ({
  name,
  fileUrl = '',
}: AdditionalFileUploadProps) => {
  const [selectedFile] = useState<string | null>(fileUrl);

  const {
    control,
    setValue,
    formState: { errors },
  } = useForm<Record<string, File>>({
    mode: 'onBlur',
    reValidateMode: 'onChange',
  });

  const handleFileChange = useCallback(
    async (e: React.ChangeEvent<HTMLInputElement>) => {
      const file = e.target.files?.[0];

      if (file) {
        setValue(name, file, { shouldValidate: true });

        //TODO: Hanlde Upload File
      }
    },
    [setValue],
  );

  return (
    <div className="flex-1 w-full">
      <FormField
        control={control}
        name={name}
        render={() => (
          <FormItem>
            <div className="relative flex items-center h-[68px]">
              <Input
                id="file-upload"
                type="file"
                className="hidden"
                onChange={handleFileChange}
                accept=".pdf"
              />
              <div className="absolute inset-0 flex items-center px-6 w-full h-full bg-blue-light rounded-[15px] test-huy">
                <span className="mr-[291px] text-md truncate text-black-smoky-">
                  {selectedFile}
                </span>
              </div>
              <Button
                variant="secondary"
                className="absolute bottom-0 right-0 w-[291px] py-[12px] h-full rounded-[15px]"
              >
                <Label
                  htmlFor="file-upload"
                  className="text-lg text-bold cursor-pointer"
                >
                  Upload
                </Label>
              </Button>
            </div>
            {errors?.file && (
              <FormMessage className="">
                <div className="flex flex-col gap-4 text-red-500 text-sm">
                  <p>{errors.file.message}</p>
                  <p>Please select again!!!</p>
                </div>
              </FormMessage>
            )}
          </FormItem>
        )}
      />
    </div>
  );
};

export default AdditionalFileUpload;
