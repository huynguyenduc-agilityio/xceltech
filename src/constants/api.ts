// TODO: Update API after BE update
const END_POINTS = {
  ACCOUNT: '/accounts/',
  LOGIN: '/accounts/login/',
  REGISTER: '/accounts/register/',
  ACTIVE: '/accounts/activate',
  USER_DETAIL: '/accounts/me/',
  LEAVES: '/leave-applications/',
  USER_LEAVES: '/leave-applications/users',
  DOCUMENT_PROFILE: '/documents/',
  NOTIFICATIONS: '/notifications/',
  JOB: '/jobs/',
};

export { END_POINTS };
