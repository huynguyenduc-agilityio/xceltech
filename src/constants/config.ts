export const REACT_QUERY = {
  STALE_TIME: 30000,
};
