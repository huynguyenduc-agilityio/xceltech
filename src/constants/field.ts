import { IEmployeeGuarantor } from '@/types';

type FormFieldKey<T> = Exclude<keyof T, 'id'>;

type BaseField<T> = {
  name: FormFieldKey<T>;
  label: string;
  placeholder: string;
};

export const BASE_GUARANTOR_FIELD: BaseField<IEmployeeGuarantor>[] = [
  {
    name: 'name',
    label: 'Guarantor’s Name',
    placeholder: 'Enter your guarantor’s name',
  },
  {
    name: 'job',
    label: 'Job Title / Occupation',
    placeholder: 'Enter your job title / occupation',
  },
  {
    name: 'phone',
    label: 'Phone No',
    placeholder: 'Enter your phone number',
  },
];
