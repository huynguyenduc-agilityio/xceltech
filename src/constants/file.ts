export enum FileType {
  Pdf = 'pdf',
  Csv = 'csv',
  Excel = 'excel',
}
