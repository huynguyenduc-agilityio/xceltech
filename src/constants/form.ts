export const LEAVE_TYPE_OPTIONS = [
  { value: 'annual leave', label: 'Annual Leave' },
  { value: 'sick', label: 'Sick' },
  { value: 'exam', label: 'Exam' },
  { value: 'maternity', label: 'Maternity' },
];

export const RELIEF_OFFICER_OPTIONS = [
  { value: 'June Heaney', label: 'June Heaney' },
  { value: 'Rudy Witting', label: 'Rudy Witting' },
  { value: 'Elaine Prohaska', label: 'Elaine Prohaska' },
  { value: 'Jordan Corwin', label: 'Jordan Corwin' },
  { value: 'Nadine Collins', label: 'Nadine Collins' },
  { value: 'Caleb Thiel', label: 'Caleb Thiel' },
];

//TODO: add more department
export const DEPARTMENT_OPTIONS = [
  { value: 'designMarketing', label: 'Design & Marketing' },
];

//TODO: add more job
export const JOB_TITLE_OPTIONS = [
  { value: 'designer', label: 'UI / UX Designer' },
];

export const JOB_CATEGORY_OPTIONS = [
  { value: 'fulltime', label: 'Full time' },
  { value: 'parttime', label: 'Part time' },
];
