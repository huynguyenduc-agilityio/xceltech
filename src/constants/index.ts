export * from './navigation';
export * from './unit';
export * from './regex';
export * from './pagination';
export * from './toast';
export * from './form';
export * from './router';
export * from './api';
export * from './config';
export * from './localStorageKey';
export * from './queries';
export * from './field';
export * from './file';
export * from './images';
export { default as MESSAGES } from './message';
