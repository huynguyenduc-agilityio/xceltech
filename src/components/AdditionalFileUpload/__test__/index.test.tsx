import { fireEvent, render } from '@testing-library/react';

// Components
import AdditionalFileUpload from '..';

// Constants
import { MESSAGES } from '@/constants';

jest.mock('react-hook-form', () => ({
  ...jest.requireActual('react-hook-form'),
  useFormContext: () => ({
    getFieldState: jest.fn(),
    formState: { errors: {} },
  }),
}));

describe('AdditionalFileUpload component', () => {
  const name = 'file';
  const mockOnFileChange = jest.fn();

  it('should match snapshot', () => {
    const { container } = render(
      <AdditionalFileUpload name={name} onFileChange={mockOnFileChange} />,
    );

    expect(container).toMatchSnapshot();
  });

  beforeEach(() => {
    global.URL.createObjectURL = jest.fn(() => 'mock-url');
    global.URL.revokeObjectURL = jest.fn();

    mockOnFileChange.mockReset();
  });

  it('renders correctly with no file', () => {
    const { getByRole } = render(
      <AdditionalFileUpload
        name={name}
        value={null}
        onFileChange={mockOnFileChange}
      />,
    );

    expect(getByRole('button', { name: /upload/i })).toBeInTheDocument();
  });

  it('handles file selection and calls onFileChange', () => {
    const { getByTitle } = render(
      <AdditionalFileUpload
        name={name}
        value={null}
        onFileChange={mockOnFileChange}
      />,
    );

    const input = getByTitle('Upload Additional File');
    const testFile = new File(['dummy content'], 'example.pdf', {
      type: 'application/pdf',
    });

    fireEvent.change(input, {
      target: { files: [testFile] },
    });

    expect(mockOnFileChange).toHaveBeenCalledWith(testFile);
  });

  it('displays the file name when a File is passed as value', () => {
    const file = new File(['file content'], 'resume.pdf', {
      type: 'application/pdf',
    });

    const { getByText, getByRole } = render(
      <AdditionalFileUpload
        name={name}
        value={file}
        onFileChange={mockOnFileChange}
      />,
    );

    const button = getByRole('button', { name: 'Upload' });

    fireEvent.click(button);

    expect(getByText('resume.pdf')).toBeInTheDocument();
  });

  it('renders link when value is a string URL', () => {
    const fileUrl = 'https://example.com/files/contract.pdf';

    const { getByRole } = render(
      <AdditionalFileUpload
        name={name}
        value={fileUrl}
        onFileChange={mockOnFileChange}
      />,
    );

    const link = getByRole('link');
    expect(link).toHaveAttribute('href', fileUrl);
    expect(link).toHaveTextContent('contract.pdf');
  });

  it('displays error message when error is present', () => {
    const { getByText } = render(
      <AdditionalFileUpload
        name={name}
        value={null}
        onFileChange={mockOnFileChange}
        errors={{
          file: {
            message: MESSAGES.VALIDATE.FIELD_REQUIRED('file'),
            type: 'required',
          },
        }}
      />,
    );

    expect(
      getByText(MESSAGES.VALIDATE.FIELD_REQUIRED('file')),
    ).toBeInTheDocument();
    expect(getByText('Please select again!!!')).toBeInTheDocument();
  });
});
