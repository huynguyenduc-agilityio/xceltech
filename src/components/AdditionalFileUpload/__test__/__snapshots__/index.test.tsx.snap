// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdditionalFileUpload component should match snapshot 1`] = `
<div>
  <div
    class="space-y-2 w-full"
  >
    <div
      class="relative flex items-center h-[68px]"
    >
      <input
        accept=".pdf"
        autocomplete="off"
        class="font-productsans w-full px-5 py-4 text-lg text-black-smoky placeholder:text-gray-soft outline-none disabled:cursor-not-allowed disabled:opacity-50 bg-blue-light h-[54px] hidden"
        data-testid="input-element"
        id="file-upload-file"
        title="Upload Additional File"
        type="file"
      />
      <div
        class="absolute inset-0 flex items-center px-6 w-full h-full bg-blue-light rounded-[15px]"
      >
        <span
          class="mr-[291px] text-md truncate text-black-smoky-"
        />
      </div>
      <button
        class="font-productsans inline-flex items-center justify-center whitespace-nowrap text-lg font-bold ring-offset-background transition-all duration-300 ease-in-out outline-none focus-visible:outline-none disabled:opacity-50 disabled:cursor-not-allowed bg-secondary text-black-default hover:bg-secondary/80 absolute bottom-0 right-0 w-[291px] py-[12px] h-full rounded-[15px]"
        type="button"
      >
        Upload
      </button>
    </div>
  </div>
</div>
`;
